### PR TITLE
Adding MetaData `@TimeIt` decorator

### LINKED ISSUE #115: Edge timings
One nice property of resource graphs and network theory is the ability to time stages. It would be quite a nice feature to add node execution timing and so-called `meta-data` to each node. This could lead to interesting analysis of the graph and possible identification of workflow failures or bottlenecks.

### Addition

This is actually an interesting step in the direction of including loops in a computational graph. Whilst the full graph should remain acyclic, it could be possible to have specific nodes in loops to solve some problem. 

What we would then have is a dynamic rather than a static graph.

### PR DESCRIPTION
**Pull request by PythonFZ**
- fix #115 

# TODOs:
- [x] set logging back to Info
- [x] tests
- [x] docstrings
- [x] documentation
- [x] should work without defining any other ZnTrackOption
- [x] should not be added as a dependency if the Node is a dependency
- [x] should probably be written to a file `Node/<name>/metadata.json`
- [ ] should have an option to select between `zn.outs` or `zn.metrics` -> zn.outs would avoid showing it with `dvc metrics show` -> moved to #123 
- [x] adding it to another method will always log the latest call to that method - consider adding a counter, if running the same method multiple times `metadata.sleep, metadata.sleep_1, metadata.sleep_2, ...`
- [ ] ~~add ~~global `config.collect_metadata` and~~ `@Node(collect_metadata: bool)`~~

# Changes + Example
- Introduce decorators based on `zntrack.metadata.MetaData` that can write custom outputs
- add `TimeIt` decorator

```py
from zntrack import Node, zn
from zntrack.metadata.decorators import TimeIt
from time import sleep

@Node()
class SleepNode:
    outs = zn.outs()
    
    @TimeIt
    def run(self):
        self.sleep_2()
        self.outs = "Lorem Ipsum"
        self.sleep_5()
    
    @TimeIt
    def sleep_5(self):
        sleep(5)
    
    @TimeIt
    def sleep_2(self):
        sleep(2)
```

`dvc metrics show`

| Path                         | metadata.run:timeit   | metadata.sleep_2:timeit   | metadata.sleep_5:timeit   |
|------------------------------|-----------------------|---------------------------|---------------------------|
| nodes/SleepNode/metrics.json | 7.00431               | 2.00063                   | 5.0028                    |

`dvc metrics diff`

| Path                         | Metric                  | HEAD    | workspace   | Change   |
|------------------------------|-------------------------|---------|-------------|----------|
| nodes/SleepNode/metrics.json | metadata.run:timeit     | 7.00431 | 7.00804     | 0.00373  |
| nodes/SleepNode/metrics.json | metadata.sleep_2:timeit | 2.00063 | 2.00213     | 0.00149  |
| nodes/SleepNode/metrics.json | metadata.sleep_5:timeit | 5.0028  | 5.00517     | 0.00237  |

```py
print(SleepNode(load=True).metadata)
> {'sleep_2:timeit': 2.002126455307007, 'sleep_5:timeit': 5.005168914794922, 'run:timeit': 7.008039236068726}

```

### CODE REVIEW

File: CI/integration_tests/dependencies/test_metadata_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    print_name = PrintName()
+
+    hello_world()
+    print_name()
+
+    project.repro()
+
+    # should only depend on the dvc.outs and not on the metric
+    assert len(set(PrintName(load=True).zntrack.dvc.deps)) == 1
+
```
Comment (PythonFZ) [inline CI/integration_tests/dependencies/test_metadata_dependencies.py:54]: This should not require a set, but the option is added twice in the dataclass

**Review by SamTov — APPROVED** [review]
A real great addition. I think that as more ideas for node metadata come up, we may need to think of a good way to handle adding the recording to nodes or elements within the node but it is a great step in that direction.

File: CI/integration_tests/dependencies/test_metadata_dependencies.py (hunk 1 above)
Comment (SamTov) [reply CI/integration_tests/dependencies/test_metadata_dependencies.py:55]: Nt sure how to get around it.

Comment (PythonFZ) [reply CI/integration_tests/dependencies/test_metadata_dependencies.py:55]: This is not urgent and does not affect the code. I opened #120 and will continue there.

**Pull request merged**